<a href="https://colab.research.google.com/github/RPGraciotti/BootCampAlura/blob/main/Modulo_6/Aula_1_Contextualiza%C3%A7%C3%A3o_do_mercado_financeiro.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sessão 1 - Contexto financeiro
Introdução:

Pilares da cadeia de crédito:

1. Prospecção
2. Concessão - Foco do módulo do BootCamp
3. Gestão de Risco 
4. Recuperação

Pedido de crédito:
Instituição financeira utiliza um modelo de decisão + política de crédito

Bytebank é o banco fictício que a Alura utiliza para ferramentas didáticas e será trabalhado durante o curso.
Streamlit é uma biblioteca python que é utilizada para hospedar o modelo e gera uma interface sem que seja necessário conhecimento de front-end.

Projeto final: Simulador de avaliação de crédito utilizando o streamlit.

Objetivo é derivado de um desafio do kaggle, traduzido e manipulado para utilização: https://www.kaggle.com/rikdifos/credit-card-approval-prediction

# Sessão 2 - Primeiros passos com a base

In [1]:
import pandas as pd

In [2]:
df_cad = pd.read_csv("https://github.com/RPGraciotti/BootCampAlura/raw/main/Data/clientes_cadastrados.txt")
df_cad

ID_Cliente Genero  ...  Tamanho_Familia  Anos_empregado
0          5008804      H  ...              2.0            12.4
1          5008805      H  ...              2.0            12.4
2          5008806      H  ...              2.0             3.1
3          5008808      M  ...              1.0             8.4
4          5008809      M  ...              1.0             8.4
...            ...    ...  ...              ...             ...
438552     6840104      H  ...              1.0         -1000.7
438553     6840222      M  ...              1.0             8.2
438554     6841878      M  ...              1.0             1.0
438555     6842765      M  ...              2.0         -1000.7
438556     6842885      M  ...              2.0             3.3

[438557 rows x 18 columns]

In [3]:
df_cad.shape

(438557, 18)

In [4]:
df_cad.head()

ID_Cliente Genero  Tem_Carro  ...   Ocupacao  Tamanho_Familia  Anos_empregado
0     5008804      H          1  ...     Outros              2.0            12.4
1     5008805      H          1  ...     Outros              2.0            12.4
2     5008806      H          1  ...  Seguranca              2.0             3.1
3     5008808      M          0  ...      Venda              1.0             8.4
4     5008809      M          0  ...      Venda              1.0             8.4

[5 rows x 18 columns]

In [5]:
df_cad.loc[:,"ID_Cliente"].value_counts().unique() # significa que só tem no máximo id com 1 repetição

array([2, 1])

In [6]:
print(df_cad.duplicated().sum())
print(df_cad.loc[:,"ID_Cliente"].duplicated(keep = False).sum())

0
94


In [7]:
df_cad[df_cad.loc[:,"ID_Cliente"].duplicated(keep = False)] # keep = True mantém só o primeiro id repetido, 
# usando keep = True mantemos todos os valores repetidos

ID_Cliente Genero  ...  Tamanho_Familia  Anos_empregado
421211     7702516      M  ...              4.0             3.4
421268     7602432      H  ...              2.0             3.6
421349     7602432      M  ...              2.0         -1000.7
421464     7836971      H  ...              3.0            15.1
421698     7213374      H  ...              2.0             2.6
...            ...    ...  ...              ...             ...
433158     7282535      M  ...              2.0         -1000.7
433159     7742853      H  ...              1.0             4.6
433217     7135270      M  ...              2.0         -1000.7
433666     7091721      M  ...              2.0             6.2
433789     7618285      M  ...              1.0             2.8

[94 rows x 18 columns]

# Desafio 1: Explorar IDs repetidos (Atualização de cadastro? Erro?)

Exemplo do cliente 7602432: Tem duas entradas, a segunda entrada tem valores muito discrepantes, levando a crer ser um erro. Mas o mais interessante é a coluna "Anos_empregado" dando um valor de -1000. Isso certamente parece um erro de preenchimento. Atentar para verificar se isso se repete.

In [8]:
ids_repetidos = df_cad[df_cad.loc[:,"ID_Cliente"].duplicated(keep = False)].loc[:,"ID_Cliente"]

In [9]:
import matplotlib.pyplot as plt

In [10]:
df_cad.iloc[ids_repetidos.index,].sort_values(by = ["ID_Cliente"]).head(20)

ID_Cliente Genero  ...  Tamanho_Familia  Anos_empregado
426818     7022197      H  ...              5.0             2.0
425023     7022197      M  ...              1.0             4.9
431545     7022327      M  ...              1.0            14.5
431911     7022327      H  ...              2.0             4.6
425486     7023108      H  ...              3.0             4.6
426488     7023108      M  ...              2.0             3.5
425306     7023651      M  ...              1.0             3.3
421907     7023651      H  ...              3.0             4.0
427778     7024111      H  ...              4.0             0.3
432643     7024111      M  ...              2.0             4.2
424708     7036518      M  ...              2.0             5.4
429078     7036518      M  ...              3.0             7.2
427007     7045794      M  ...              1.0             4.0
428048     7045794      M  ...              4.0             1.4
432271     7045885      M  ...              4.0             4.4
426022     7045885      H  ...              2.0         -1000.7
432134     7046068      M  ...              1.0             7.6
424618     7046068      M  ...              2.0         -1000.7
428171     7050948      M  ...              2.0            11.0
430746     7050948      M  ...              2.0             0.4

[20 rows x 18 columns]

In [11]:
df_cad_no_rep = df_cad.drop(ids_repetidos.index)

In [12]:
df_cad_no_rep.loc[:,"ID_Cliente"].value_counts()

6293503    1
5166449    1
6210931    1
5174645    1
5170551    1
          ..
7410037    1
6357367    1
6375806    1
5683584    1
6291456    1
Name: ID_Cliente, Length: 438463, dtype: int64

In [13]:
df_cad_no_rep.isnull().sum()

ID_Cliente               0
Genero                   0
Tem_Carro                0
Tem_Casa_Propria         0
Qtd_Filhos               0
Rendimento_Anual         0
Categoria_de_renda       0
Grau_Escolaridade        0
Estado_Civil             0
Moradia                  0
Idade                    0
Tem_Celular              0
Tem_telefone_trabalho    0
Tem_telefone_fixo        0
Tem_email                0
Ocupacao                 0
Tamanho_Familia          0
Anos_empregado           0
dtype: int64

In [14]:
df_cad_no_rep.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 438463 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   ID_Cliente             438463 non-null  int64  
 1   Genero                 438463 non-null  object 
 2   Tem_Carro              438463 non-null  int64  
 3   Tem_Casa_Propria       438463 non-null  int64  
 4   Qtd_Filhos             438463 non-null  int64  
 5   Rendimento_Anual       438463 non-null  float64
 6   Categoria_de_renda     438463 non-null  object 
 7   Grau_Escolaridade      438463 non-null  object 
 8   Estado_Civil           438463 non-null  object 
 9   Moradia                438463 non-null  object 
 10  Idade                  438463 non-null  float64
 11  Tem_Celular            438463 non-null  int64  
 12  Tem_telefone_trabalho  438463 non-null  int64  
 13  Tem_telefone_fixo      438463 non-null  int64  
 14  Tem_email              438463 non-nu

In [15]:
df_cad_no_rep.describe()

ID_Cliente      Tem_Carro  ...  Tamanho_Familia  Anos_empregado
count  4.384630e+05  438463.000000  ...    438463.000000   438463.000000
mean   6.021894e+06       0.371906  ...         2.194461     -165.936036
std    5.713554e+05       0.483314  ...         0.897194      380.201131
min    5.008804e+06       0.000000  ...         1.000000    -1000.700000
25%    5.609340e+06       0.000000  ...         2.000000        1.000000
50%    6.047690e+06       0.000000  ...         2.000000        4.000000
75%    6.454118e+06       1.000000  ...         3.000000        8.500000
max    7.999952e+06       1.000000  ...        20.000000       48.000000

[8 rows x 12 columns]

O valor de -1000 "anos empregado" também aparece no conjunto sem os dados repetidos. Provavelmente lidaremos com esse erro de novo mais para frente. Comparando os padrões dos clientes com IDs repetidos, não é possível determinar qual a origem do erro: são identidades completamente diferentes mesmo, pessoas com estados civis diferentes e tamanho familiar diferente (uma hipótese é que fossem casais). A princípio não há como inferir a fonte do erro. A minha hipótese é de que um número de cadastro ficou inválido (a pessoa fechou a conta no banco, por exemplo) e utilizaram o mesmo ID para um novo cliente, sem remover os antigos da base. Dessa forma, não é possível determinar qual o cliente "novo" e o cliente "antigo".

# Desafio 2: Explorar as colunas (features), identificar contínuas, categóricas e binárias

In [16]:
df_cad_no_rep.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 438463 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   ID_Cliente             438463 non-null  int64  
 1   Genero                 438463 non-null  object 
 2   Tem_Carro              438463 non-null  int64  
 3   Tem_Casa_Propria       438463 non-null  int64  
 4   Qtd_Filhos             438463 non-null  int64  
 5   Rendimento_Anual       438463 non-null  float64
 6   Categoria_de_renda     438463 non-null  object 
 7   Grau_Escolaridade      438463 non-null  object 
 8   Estado_Civil           438463 non-null  object 
 9   Moradia                438463 non-null  object 
 10  Idade                  438463 non-null  float64
 11  Tem_Celular            438463 non-null  int64  
 12  Tem_telefone_trabalho  438463 non-null  int64  
 13  Tem_telefone_fixo      438463 non-null  int64  
 14  Tem_email              438463 non-nu

In [17]:
df_cad_no_rep.describe()

ID_Cliente      Tem_Carro  ...  Tamanho_Familia  Anos_empregado
count  4.384630e+05  438463.000000  ...    438463.000000   438463.000000
mean   6.021894e+06       0.371906  ...         2.194461     -165.936036
std    5.713554e+05       0.483314  ...         0.897194      380.201131
min    5.008804e+06       0.000000  ...         1.000000    -1000.700000
25%    5.609340e+06       0.000000  ...         2.000000        1.000000
50%    6.047690e+06       0.000000  ...         2.000000        4.000000
75%    6.454118e+06       1.000000  ...         3.000000        8.500000
max    7.999952e+06       1.000000  ...        20.000000       48.000000

[8 rows x 12 columns]

In [18]:
df_cad_no_rep.head(2)

ID_Cliente Genero  Tem_Carro  ...  Ocupacao  Tamanho_Familia  Anos_empregado
0     5008804      H          1  ...    Outros              2.0            12.4
1     5008805      H          1  ...    Outros              2.0            12.4

[2 rows x 18 columns]

Features contínuas: "Rendimento_Anual", "Idade", "Tamanho Família", "Anos empregado", "Quantidade de filhos".

Features categóricas: "Categoria de renda", "Grau escolaridade", "Estado civil", "Moradia", "Ocupação".

Features categóricas binárias: "Genero", "Tem carro", "Tem casa propria", "Tem celular", "Tem telefone trabalho", "Tem telefone fixo", "Tem email".

In [19]:
cont = ["Rendimento_Anual", "Idade", "Tamanho_Familia", "Anos_empregado", "Qtd_filhos"]
cat = ["Categoria_de_renda", "Grau_Escolaridade", "Estado_Civil", "Moradia", "Ocupacao"]
bin = ["Genero", "Tem_Carro", "Tem_Casa_Propria", "Tem_Celular", "Tem_telefone_trabalho", "Tem_telefone_fixo", "Tem_email"]